proto/leaderboard.proto

```diff
@@ -0,0 +1,192 @@
+  rpc GetRankAsc(GetRankRequest) returns (RankedElement);
+
+  // Get the rank of a particular id in the leaderboard.
+  // * Ordinal, 0-based rank.
+  // * Descending order (0 is the highest-scoring rank).
+  rpc GetRankDesc(GetRankRequest) returns (RankedElement);
+
+  // Get a range of elements by a score range.
+  // * Ordinal, 0-based rank.
+  // * Ascending order (0 is the lowest-scoring rank)
+  //
+  // You can request up to 8192 elements at a time. To page through many elements that all
+  // fall into a score range you can repeatedly invoke this api with the offset parameter.
+  rpc GetByScoreAsc(GetByScoreRequest) returns (GetByScoreResponse);
+
+  // Get a range of elements by a score range.
+  // * Ordinal, 0-based rank.
+  // * Descending order (0 is the highest-scoring rank)
+  //
+  // You can request up to 8192 elements at a time. To page through many elements that all
+  // fall into a score range you can repeatedly invoke this api with the offset parameter.
+  rpc GetByScoreDesc(GetByScoreRequest) returns (GetByScoreResponse);
+}
+
+// Leaderboards are made up of many of these.
+message Element {
+  // A player identifier, session identifier, browser identifier or whatever other kind of
+  // identifier you use for this scoreboard. The full unsigned 64 bit range is allowed here,
+  // between 0 and 2^63-1 inclusive.
+  // An id can only appear in a leaderboard one time. You can't have 2 scores for 1 player,
+  // unless that player has 2 ids!
+  uint64 id = 1;
+
+  // The value by which this element is sorted within the leaderboard.
+  float score = 2;
+}
+
+// Query APIs returning RankedElement offer the familiar Element id and score tuple, but they
+// also include the rank per the individual API's ranking semantic.
+message RankedElement {
+  uint64 id = 1;
+  float score = 2;
+  uint64 rank = 3;
+}
+
+// Query APIs using RankRange expect a limit of 8192 elements. Requesting a range wider than
+// that is expected to return an error.
+//
+// RankRange models half-open ranges: 0..4 refers to elements 0, 1, 2 and 3.
+//
+// Example permissible ranges:
+// * 0..8192
+// * 1..8193
+// * 123..8315
+// * 0..1
+// * 1..4
+// * 13..17
+//
+// Example error ranges:
+// * 0..0
+// * 4..3
+// * 0..8193
+message RankRange {
+  uint64 start_inclusive = 1;
+  uint64 end_exclusive = 2;
+}
+
+// Query APIs using ScoreRange may match more than the limit of 8192 elements. These apis will
+// include an offset and limit parameter pair, which can be used to page through score ranges
+// matching many elements.
+//
+// ScoreRange models half-open ranges: 0..4 refers to scores 0, 1.1234, 2.5 and 3.999.
+message ScoreRange {
+  // IEEE 754 single precision 32 bit floating point number.
+  // Momento does not support NaN or Inf in leaderboards.
+  float score_start = 1;
+  // IEEE 754 single precision 32 bit floating point number.
+  // Momento does not support NaN or Inf in leaderboards.
+  float score_end = 2;
+}
+
+message Empty {}
```

Comment:
@kvcache have you considered using GRPC's default `Empty` message? That way we don't have to have multiple classpaths for an empty message

Reply:
Yes I did consider that. But this is easier and doesn't have any extra import dependencies. It's also binary-compatible with that message if you ever want to delete this line and replace it with another line that imports this exact structure from someone else's code :-p! This is basically just a more intent-revealing replacement for `XRpcResponse {}` boilerplate that we copypaste and replace X for in many other places.